network/p2p/node_sampler.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	safemath "github.com/ava-labs/avalanchego/utils/math"
+	"github.com/ava-labs/avalanchego/utils/sampler"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/version"
+)
+
+const (
+	maxValidatorSetStaleness = time.Minute
+)
+
+var (
+	_ NodeSampler = (*Validators)(nil)
+
+	_ validators.Connector = (*Peers)(nil)
+	_ NodeSampler          = (*Peers)(nil)
+)
+
+type NodeSampler interface {
+	// Sample returns at most [n] nodes. This may return fewer nodes if fewer
+	// than [n] are available.
+	Sample(n int) ([]ids.NodeID, error)
```

Comment:
Can you give an example of a sample that would return an error?

Reply:
> Shouldn't Sample take a context? You are already needing to stub out TODO getRecentValidatorIDs.

Yeah this makes sense, I can add this to the signature

> Can you give an example of a sample that would return an error?

Hmm I guess an example would be some grpc failure? Some implementation might hit the p-chain for example

Reply:
Actually, the only reason we have this `Sample` interface is to use it for the `Client` node selection implementation. We could actually swallow this error and just return an empty slice of node ids, and have the `Client` report an error if it sees no nodes to send to, so the error is still propagated and this interface signature is much cleaner since we can get rid of the error here.

Reply:
> We could actually swallow this error and just return an empty slice of node ids, and have the Client report an error if it sees no nodes to send to,

If we error and return on calls to `v.validators.GetCurrentHeight(ctx)` or `v.validators.GetValidatorSet(ctx, height, v.subnetID)` we won't update `v.validatorIDs` so future sampling won't return an empty slice of node IDs, it'll just return stale node IDs. FWIW the sampler implementations in the `sampler` package have errors in their return signature.

Reply:
I added a unit test for this specific scenario to make sure we return an empty slice instead of stale ids so the error is surfaced correctly in the `Client`.